x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/sort/DoubleBucketedSort.java

```diff
@@ -0,0 +1,307 @@
+    private DoubleArray values;
+
+    public DoubleBucketedSort(BigArrays bigArrays, SortOrder order, int bucketSize) {
+        this.bigArrays = bigArrays;
+        this.order = order;
+        this.bucketSize = bucketSize;
+        heapMode = new BitArray(0, bigArrays);
+
+        boolean success = false;
+        try {
+            values = bigArrays.newDoubleArray(0, false);
+            success = true;
+        } finally {
+            if (success == false) {
+                close();
+            }
+        }
+    }
+
+    /**
+     * The order of the sort.
+     */
+    public SortOrder getOrder() {
+        return order;
+    }
+
+    /**
+     * The number of values to store per bucket.
+     */
+    public int getBucketSize() {
+        return bucketSize;
+    }
+
+    /**
+     * Get the bound indexes (inclusive, exclusive) of the values for a bucket.
+     * Returns [0, 0] if the bucket has never been collected.
+     */
+    private Tuple<Long, Long> getBucketValuesBounds(long bucket) {
+        long rootIndex = bucket * bucketSize;
+        if (rootIndex >= values.size()) {
+            // We've never seen this bucket.
+            return Tuple.tuple(0L, 0L);
+        }
+        long start = inHeapMode(bucket) ? rootIndex : (rootIndex + getNextGatherOffset(rootIndex) + 1);
+        long end = rootIndex + bucketSize;
+        return Tuple.tuple(start, end);
+    }
+
+    /**
+     * Merge the values from {@code other}'s {@code otherGroupId} into {@code groupId}.
+     */
+    public void merge(int groupId, DoubleBucketedSort other, int otherGroupId) {
+        var otherBounds = other.getBucketValuesBounds(otherGroupId);
+
+        // TODO: This can be improved for heapified buckets by making use of the heap structures
+        for (long i = otherBounds.v1(); i < otherBounds.v2(); i++) {
+            collect(other.values.get(i), groupId);
+        }
+    }
+
+    /**
+     * Creates a block with the values from the {@code selected} groups.
+     */
+    public Block toBlock(BlockFactory blockFactory, IntVector selected) {
+        // Used to sort the values in the bucket.
+        var bucketValues = new double[bucketSize];
+
+        try (var builder = blockFactory.newDoubleBlockBuilder(selected.getPositionCount())) {
+            for (int s = 0; s < selected.getPositionCount(); s++) {
+                int bucket = selected.getInt(s);
+
+                var bounds = getBucketValuesBounds(bucket);
+                var size = bounds.v2() - bounds.v1();
+
+                if (size == 0) {
+                    builder.appendNull();
+                    continue;
+                }
+
+                if (size == 1) {
+                    builder.appendDouble(values.get(bounds.v1()));
+                    continue;
+                }
+
+                for (int i = 0; i < size; i++) {
+                    bucketValues[i] = values.get(bounds.v1() + i);
+                }
+
+                Arrays.sort(bucketValues, 0, (int) size);
```

Comment:
And `downHeap`. It might be worth adding a comment to those two methods saying "these are on the hot path for aggs, we are allocating as little as possible"

Reply:
I was checking the topic, but we can't iterate a min heap in order, we have to copy and sort it (?)
That's why I discarded the comment, but I may be wrong

Reply:
> I was checking the topic, but we can't iterate a min heap in order, we have to copy and sort it (?)
> That's why I discarded the comment, but I may be wrong

It's a min-heap so we could make something to do that iteration. It's finicky code that'd want its own unit test, probably randomized. But we totally could do it. If you leave the TODO I'll try and have a look. It may not be worth it. But it feels bad to ignore the heap property when sorting.